### ISSUE TITLE
Mypy doesn't respect my assert statement

### ISSUE THREAD

**Issue by ThatXliner** [python/mypy#10186]
**Bug Report**

For an optional type, mypy fails to respect my explicit assert statement

(A clear and concise description of what the bug is.)

**To Reproduce**
MyPy lint this (strict or no strict, still the same error):
```py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Git utilities"""
import shutil
import subprocess
from typing import Optional, List

GIT = shutil.which("git")
if GIT is None:
    raise RuntimeError("Git is not installed on this system.")
assert GIT is not None  # Should shut up mypy


def get_lst_commit_msg(branch: Optional[str] = None) -> str:
    command: List[str] = [GIT, "log", "-1", "--pretty=%B"]  # Mypy complains here ;-;
    if branch:
        command.append(branch)
    return subprocess.check_output(command).decode()
```

**Expected Behavior**
It shouldn't complain

**Actual Behavior**

```http
foo.py:15: error: List item 0 has incompatible type "Optional[str]"; expected "str"
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: mypy 0.812
- Mypy command-line flags: `path/to/foo.py` and/or `--strict` 
- Mypy configuration options from `mypy.ini` (and other config files): None that I know of
- Python version used: Python 3.6.12
- Operating system and version: `MacOS 10.15.7`

**Comment by hauntsaninja** [comment]
The type of the variable is still Optional, since you could reassign to GIT later

**Comment by ThatXliner** [comment]
Woah! I didn't think how fast the feedback would be 😄 ! Thanks for the help! I didn't think of that 😅
